File: simulateur.py
"""
Simulateur de Gestion Aérienne - Programme Principal

LOGIQUE DU SYSTÈME:

1. ARCHITECTURE:
   - Avion: Représente un avion avec son état (en approche, atterrissage, au sol, décollage)
   - Piste: Gère la disponibilité et l'occupation des pistes
   - Aéroport: Coordonne l'ensemble du trafic aérien

2. FLUX D'ATTERRISSAGE:
   a) Un avion signale son approche → ajouté à la file d'attente
   b) Le contrôle aérien assigne une piste disponible
   c) L'avion atterrit → occupe temporairement la piste
   d) L'avion libère la piste → se positionne au sol

3. FLUX DE DÉCOLLAGE:
   a) Un avion au sol demande à décoller
   b) Le contrôle aérien assigne une piste disponible
   c) L'avion décolle → occupe temporairement la piste
   d) L'avion libère la piste → quitte l'espace aérien

4. PRIORITÉS:
   - Les atterrissages sont prioritaires (sécurité)
   - File d'attente FIFO (First In, First Out)
   - Attribution automatique des pistes disponibles

Ce programme démontre le fonctionnement du système avec plusieurs scénarios.
"""

from avion import Avion
from aeroport import Aeroport
import time


def pause(secondes=1):
    """Pause pour simuler le temps réel"""
    time.sleep(secondes)


def demonstration_simple():
    """Démonstration simple du système"""
    print("\n" + "🌟"*30)
    print("DÉMONSTRATION SIMPLE DU SIMULATEUR DE GESTION AÉRIENNE")
    print("🌟"*30 + "\n")
    
    # Créer l'aéroport
    aeroport = Aeroport("Charles de Gaulle", nombre_pistes=2)
    print(f"✅ Aéroport '{aeroport.nom}' créé avec {len(aeroport.pistes)} pistes\n")
    pause()
    
    # Créer des avions
    avion1 = Avion("AF123", "Air France")
    avion2 = Avion("LH456", "Lufthansa")
    avion3 = Avion("BA789", "British Airways")
    
    print("📝 SCÉNARIO 1: Arrivée de plusieurs avions")
    print("-" * 50)
    aeroport.ajouter_avion_en_approche(avion1)
    pause()
    aeroport.ajouter_avion_en_approche(avion2)
    pause()
    aeroport.ajouter_avion_en_approche(avion3)
    pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 SCÉNARIO 2: Atterrissages")
    print("-" * 50)
    aeroport.autoriser_atterrissage()  # AF123
    pause()
    aeroport.autoriser_atterrissage()  # LH456
    pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 SCÉNARIO 3: Finalisation des atterrissages")
    print("-" * 50)
    aeroport.finaliser_atterrissage("AF123")
    pause()
    aeroport.finaliser_atterrissage("LH456")
    pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 SCÉNARIO 4: Troisième atterrissage")
    print("-" * 50)
    aeroport.autoriser_atterrissage()  # BA789
    pause()
    aeroport.finaliser_atterrissage("BA789")
    pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 SCÉNARIO 5: Décollages")
    print("-" * 50)
    aeroport.autoriser_decollage("AF123")
    pause()
    aeroport.finaliser_decollage("AF123")
    pause()
    
    aeroport.autoriser_decollage("LH456")
    pause()
    aeroport.finaliser_decollage("LH456")
    pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    # Afficher l'historique complet
    aeroport.afficher_historique()


def demonstration_complexe():
    """Démonstration avec gestion de la saturation des pistes"""
    print("\n" + "🌟"*30)
    print("DÉMONSTRATION AVANCÉE: GESTION DE LA SATURATION")
    print("🌟"*30 + "\n")
    
    # Aéroport avec une seule piste
    aeroport = Aeroport("Orly", nombre_pistes=1)
    print(f"✅ Aéroport '{aeroport.nom}' créé avec {len(aeroport.pistes)} piste\n")
    pause()
    
    # Plusieurs avions arrivent simultanément
    avions = [
        Avion("EZ101", "EasyJet"),
        Avion("RY202", "Ryanair"),
        Avion("VY303", "Vueling")
    ]
    
    print("📝 SCÉNARIO: Saturation avec une seule piste")
    print("-" * 50)
    
    for avion in avions:
        aeroport.ajouter_avion_en_approche(avion)
        pause()
    
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 Premier avion atterrit (piste occupée)")
    print("-" * 50)
    aeroport.autoriser_atterrissage()
    pause()
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 Tentative d'atterrissage du deuxième (piste occupée)")
    print("-" * 50)
    aeroport.autoriser_atterrissage()  # Va échouer car piste occupée
    pause()
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 Libération de la piste")
    print("-" * 50)
    aeroport.finaliser_atterrissage("EZ101")
    pause()
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 Maintenant le deuxième peut atterrir")
    print("-" * 50)
    aeroport.autoriser_atterrissage()
    pause()
    aeroport.finaliser_atterrissage("RY202")
    pause()
    aeroport.afficher_etat()
    pause(2)
    
    print("\n📝 Et le troisième")
    print("-" * 50)
    aeroport.autoriser_atterrissage()
    pause()
    aeroport.finaliser_atterrissage("VY303")
    pause()
    
    aeroport.afficher_etat()
    aeroport.afficher_historique()


def menu_interactif():
    """Menu interactif pour contrôler manuellement l'aéroport"""
    print("\n" + "🌟"*30)
    print("MODE INTERACTIF - CONTRÔLE AÉRIEN")
    print("🌟"*30 + "\n")
    
    nom_aeroport = input("Nom de l'aéroport: ").strip() or "Mon Aéroport"
    try:
        nb_pistes = int(input("Nombre de pistes (1-5): ").strip() or "3")
        nb_pistes = max(1, min(5, nb_pistes))
    except ValueError:
        nb_pistes = 3
    
    aeroport = Aeroport(nom_aeroport, nb_pistes)
    print(f"\n✅ Aéroport '{aeroport.nom}' créé avec {len(aeroport.pistes)} piste(s)\n")
    
    while True:
        print("\n" + "="*60)
        print("MENU PRINCIPAL")
        print("="*60)
        print("1. Ajouter un avion en approche")
        print("2. Autoriser un atterrissage")
        print("3. Finaliser un atterrissage")
        print("4. Autoriser un décollage")
        print("5. Finaliser un décollage")
        print("6. Afficher l'état de l'aéroport")
        print("7. Afficher l'historique")
        print("8. Quitter")
        print("="*60)
        
        choix = input("\nVotre choix (1-8): ").strip()
        
        if choix == "1":
            code_vol = input("Code du vol (ex: AF123): ").strip().upper()
            compagnie = input("Compagnie aérienne: ").strip()
            if code_vol and compagnie:
                avion = Avion(code_vol, compagnie)
                aeroport.ajouter_avion_en_approche(avion)
            else:
                print("⚠️  Code de vol et compagnie requis")
        
        elif choix == "2":
            aeroport.autoriser_atterrissage()
        
        elif choix == "3":
            code_vol = input("Code du vol à finaliser: ").strip().upper()
            if code_vol:
                aeroport.finaliser_atterrissage(code_vol)
            else:
                print("⚠️  Code de vol requis")
        
        elif choix == "4":
            code_vol = input("Code du vol à faire décoller: ").strip().upper()
            if code_vol:
                aeroport.autoriser_decollage(code_vol)
            else:
                print("⚠️  Code de vol requis")
        
        elif choix == "5":
            code_vol = input("Code du vol à finaliser: ").strip().upper()
            if code_vol:
                aeroport.finaliser_decollage(code_vol)
            else:
                print("⚠️  Code de vol requis")
        
        elif choix == "6":
            aeroport.afficher_etat()
        
        elif choix == "7":
            aeroport.afficher_historique()
        
        elif choix == "8":
            print("\n👋 Au revoir! Merci d'avoir utilisé le simulateur.\n")
            break
        
        else:
            print("⚠️  Choix invalide, veuillez réessayer")


def main():
    """Programme principal"""
    print("\n" + "✈️ "*30)
    print(" "*20 + "SIMULATEUR DE GESTION AÉRIENNE")
    print("✈️ "*30 + "\n")
    
    print("Choisissez un mode:")
    print("1. Démonstration simple")
    print("2. Démonstration avancée (saturation)")
    print("3. Mode interactif")
    print("4. Toutes les démonstrations")
    
    choix = input("\nVotre choix (1-4): ").strip()
    
    if choix == "1":
        demonstration_simple()
    elif choix == "2":
        demonstration_complexe()
    elif choix == "3":
        menu_interactif()
    elif choix == "4":
        demonstration_simple()
        pause(3)
        demonstration_complexe()
        pause(3)
        reponse = input("\nVoulez-vous essayer le mode interactif? (o/n): ").strip().lower()
        if reponse == "o":
            menu_interactif()
    else:
        print("Choix invalide, lancement de la démonstration simple...")
        demonstration_simple()


if __name__ == "__main__":
    main()

File: java/src/Simulateur.java
import java.util.Scanner;

/**
 * Simulateur - Point d'entrée du programme Java.
 *
 * LOGIQUE DU PROGRAMME:
 *  Le programme propose trois modes :
 *   1. Démonstration simple   – scénario prédéfini avec 2 pistes
 *   2. Démonstration avancée  – gestion de la saturation (1 piste)
 *   3. Mode interactif        – l'utilisateur contrôle tout via un menu
 *
 * Lancement :
 *   javac -d out src/*.java
 *   java -cp out Simulateur
 */
public class Simulateur {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println();
        printRepeat("✈ ", 30);
        System.out.println("          SIMULATEUR DE GESTION AERIENNE (Java)");
        printRepeat("✈ ", 30);
        System.out.println();

        System.out.println("Choisissez un mode:");
        System.out.println("1. Démonstration simple");
        System.out.println("2. Démonstration avancée (saturation)");
        System.out.println("3. Mode interactif");
        System.out.print("\nVotre choix (1-3): ");

        String choix = sc.nextLine().trim();

        switch (choix) {
            case "1" -> demonstrationSimple();
            case "2" -> demonstrationComplexe();
            case "3" -> menuInteractif(sc);
            default  -> {
                System.out.println("Choix invalide, lancement de la démonstration simple...");
                demonstrationSimple();
            }
        }

        sc.close();
    }

    // ------------------------------------------------------------------ //
    //  Démonstrations prédéfinies
    // ------------------------------------------------------------------ //

    /**
     * Démonstration simple : 3 avions, 2 pistes, atterrissages puis décollages.
     *
     * OBJECTIF PÉDAGOGIQUE:
     *  Montre le cycle complet et l'effet de la file d'attente quand
     *  le nombre d'avions dépasse le nombre de pistes disponibles.
     */
    static void demonstrationSimple() {
        printStar("DÉMONSTRATION SIMPLE DU SIMULATEUR DE GESTION AÉRIENNE");

        Aeroport aeroport = new Aeroport("Charles de Gaulle", 2);
        System.out.println("Aéroport '" + aeroport.getNom() + "' créé avec "
                + aeroport.getPistes().size() + " pistes\n");

        Avion avion1 = new Avion("AF123", "Air France");
        Avion avion2 = new Avion("LH456", "Lufthansa");
        Avion avion3 = new Avion("BA789", "British Airways");

        // --- Scénario 1 : arrivée ---
        printScenario("SCÉNARIO 1 : Arrivée de plusieurs avions");
        aeroport.ajouterAvionEnApproche(avion1);
        aeroport.ajouterAvionEnApproche(avion2);
        aeroport.ajouterAvionEnApproche(avion3);
        aeroport.afficherEtat();

        // --- Scénario 2 : atterrissages ---
        printScenario("SCÉNARIO 2 : Atterrissages");
        aeroport.autoriserAtterrissage();   // AF123
        aeroport.autoriserAtterrissage();   // LH456
        aeroport.afficherEtat();

        // --- Scénario 3 : finalisation ---
        printScenario("SCÉNARIO 3 : Finalisation des atterrissages");
        aeroport.finaliserAtterrissage("AF123");
        aeroport.finaliserAtterrissage("LH456");
        aeroport.afficherEtat();

        // --- Scénario 4 : troisième avion ---
        printScenario("SCÉNARIO 4 : Troisième atterrissage");
        aeroport.autoriserAtterrissage();   // BA789
        aeroport.finaliserAtterrissage("BA789");
        aeroport.afficherEtat();

        // --- Scénario 5 : décollages ---
        printScenario("SCÉNARIO 5 : Décollages");
        aeroport.autoriserDecollage("AF123");
        aeroport.finaliserDecollage("AF123");
        aeroport.autoriserDecollage("LH456");
        aeroport.finaliserDecollage("LH456");
        aeroport.afficherEtat();

        aeroport.afficherHistorique();
    }

    /**
     * Démonstration avancée : 3 avions arrivent sur un aéroport n'ayant qu'une seule piste.
     *
     * OBJECTIF PÉDAGOGIQUE:
     *  Montre clairement que la file d'attente prend en charge la saturation :
     *  le troisième avion doit attendre la libération de la piste unique.
     */
    static void demonstrationComplexe() {
        printStar("DÉMONSTRATION AVANCÉE : GESTION DE LA SATURATION");

        Aeroport aeroport = new Aeroport("Orly", 1);
        System.out.println("Aéroport '" + aeroport.getNom() + "' créé avec "
                + aeroport.getPistes().size() + " piste\n");

        Avion[] avions = {
            new Avion("EZ101", "EasyJet"),
            new Avion("RY202", "Ryanair"),
            new Avion("VY303", "Vueling")
        };

        printScenario("Saturation avec une seule piste");
        for (Avion a : avions) {
            aeroport.ajouterAvionEnApproche(a);
        }
        aeroport.afficherEtat();

        printScenario("Premier avion atterrit");
        aeroport.autoriserAtterrissage();
        aeroport.afficherEtat();

        printScenario("Tentative du deuxième (piste occupée)");
        aeroport.autoriserAtterrissage();   // doit échouer
        aeroport.afficherEtat();

        printScenario("Libération de la piste");
        aeroport.finaliserAtterrissage("EZ101");
        aeroport.afficherEtat();

        printScenario("Maintenant le deuxième peut atterrir");
        aeroport.autoriserAtterrissage();
        aeroport.finaliserAtterrissage("RY202");
        aeroport.afficherEtat();

        printScenario("Et le troisième");
        aeroport.autoriserAtterrissage();
        aeroport.finaliserAtterrissage("VY303");
        aeroport.afficherEtat();

        aeroport.afficherHistorique();
    }

    // ------------------------------------------------------------------ //
    //  Mode interactif
    // ------------------------------------------------------------------ //

    /**
     * Menu interactif permettant de contrôler l'aéroport opération par opération.
     *
     * @param sc Scanner partagé avec main()
     */
    static void menuInteractif(Scanner sc) {
        printStar("MODE INTERACTIF - CONTRÔLE AÉRIEN");

        System.out.print("Nom de l'aéroport: ");
        String nom = sc.nextLine().trim();
        if (nom.isEmpty()) nom = "Mon Aéroport";

        int nbPistes = 3;
        System.out.print("Nombre de pistes (1-5): ");
        try {
            nbPistes = Integer.parseInt(sc.nextLine().trim());
            nbPistes = Math.max(1, Math.min(5, nbPistes));
        } catch (NumberFormatException ignored) { }

        Aeroport aeroport = new Aeroport(nom, nbPistes);
        System.out.println("\nAéroport '" + aeroport.getNom() + "' créé avec "
                + aeroport.getPistes().size() + " piste(s)\n");

        boolean continuer = true;
        while (continuer) {
            String sep = "=".repeat(60);
            System.out.println("\n" + sep);
            System.out.println("MENU PRINCIPAL");
            System.out.println(sep);
            System.out.println("1. Ajouter un avion en approche");
            System.out.println("2. Autoriser un atterrissage");
            System.out.println("3. Finaliser un atterrissage");
            System.out.println("4. Autoriser un décollage");
            System.out.println("5. Finaliser un décollage");
            System.out.println("6. Afficher l'état de l'aéroport");
            System.out.println("7. Afficher l'historique");
            System.out.println("8. Quitter");
            System.out.println(sep);
            System.out.print("\nVotre choix (1-8): ");

            String choix = sc.nextLine().trim();

            switch (choix) {
                case "1" -> {
                    System.out.print("Code du vol (ex: AF123): ");
                    String codeVol = sc.nextLine().trim().toUpperCase();
                    System.out.print("Compagnie aérienne: ");
                    String compagnie = sc.nextLine().trim();
                    if (!codeVol.isEmpty() && !compagnie.isEmpty()) {
                        aeroport.ajouterAvionEnApproche(new Avion(codeVol, compagnie));
                    } else {
                        System.out.println("⚠  Code de vol et compagnie requis");
                    }
                }
                case "2" -> aeroport.autoriserAtterrissage();
                case "3" -> {
                    System.out.print("Code du vol à finaliser: ");
                    String codeVol = sc.nextLine().trim().toUpperCase();
                    if (!codeVol.isEmpty()) {
                        aeroport.finaliserAtterrissage(codeVol);
                    } else {
                        System.out.println("⚠  Code de vol requis");
                    }
                }
                case "4" -> {
                    System.out.print("Code du vol à faire décoller: ");
                    String codeVol = sc.nextLine().trim().toUpperCase();
                    if (!codeVol.isEmpty()) {
                        aeroport.autoriserDecollage(codeVol);
                    } else {
                        System.out.println("⚠  Code de vol requis");
                    }
                }
                case "5" -> {
                    System.out.print("Code du vol à finaliser: ");
                    String codeVol = sc.nextLine().trim().toUpperCase();
                    if (!codeVol.isEmpty()) {
                        aeroport.finaliserDecollage(codeVol);
                    } else {
                        System.out.println("⚠  Code de vol requis");
                    }
                }
                case "6" -> aeroport.afficherEtat();
                case "7" -> aeroport.afficherHistorique();
                case "8" -> {
                    System.out.println("\nAu revoir! Merci d'avoir utilisé le simulateur.\n");
                    continuer = false;
                }
                default -> System.out.println("⚠  Choix invalide, veuillez réessayer");
            }
        }
    }

    // ------------------------------------------------------------------ //
    //  Utilitaires d'affichage
    // ------------------------------------------------------------------ //

    private static void printRepeat(String s, int n) {
        for (int i = 0; i < n; i++) System.out.print(s);
        System.out.println();
    }

    private static void printStar(String titre) {
        System.out.println();
        printRepeat("* ", 30);
        System.out.println(titre);
        printRepeat("* ", 30);
        System.out.println();
    }

    static void printScenario(String titre) {
        System.out.println("\n>>> " + titre);
        System.out.println("-".repeat(50));
    }
}
